query/query.q

```diff
@@ -0,0 +1,51 @@
+///
+// Helper function to flip table rows presented as dictionaries.
+// Flip extends atoms automagically, but strings are treated
+//  as compound lists of characters which won't flip due to
+//  inconsistent length with other compound columns.
+// @param strCols List of symbols specifying the columns to be treated as strings.
+// @param d Dictionary of column vectors.
+// @return Flipped row dictionary to be used as a segment of new table.
+.finos.query.xflipDict:{[strCols;d]
+  if[99h<>type d; '"d must be a dictionary from a table row"];
+  if[11h<>abs type strCols; '"strCols must be a symbol or list thereof"];
+  strCols,:();         / Ensure strCols is a list.
+  nsc:strCols _ d;     / Non-string column vectors.
+  errMsg:"Failed to flip non-string columns. "
+        ,"Check that all string columns were specified.  Error: ";
+  // Flip the non-string column vectors into a table.
+  r:@[flip;nsc;{[errMsg;err]'errMsg,-3!err}[errMsg;]];
+  // Replicate string values to match length of r.
+  sc:strCols!(count[r]#enlist@)each value strCols#d;
+  // Then put it back into a table structure and paste to r.
+  // Put columns back in original order.
+  key[d]xcols r,'flip sc
+ }
+
+///
+// Treat columns named in "strCols" as strings that are not
+//  to be ungrouped.  Useful for tables that have strings for
+//  GUID columns.
+// @param strCols List of symbols specifying the columns to be treated as strings.
+// @param t Table to be ungrouped.
+// @return Table that has had nested subvectors flattened out into
+//           multiple rows.
+.finos.query.xungroup:{[strCols;t]
+  if[`~strCols; : ungroup t];
+  $[count t:()xkey t
+   ;raze .finos.query.xflipDict[strCols;]each t
+   ;t]
+ }
+
+
+///
+// Ungroup a table that might use vectors of characters as strings.
+// Use .finos.query.xungroup[...] for finer control.
+// @param t Table to be ungrouped.
+.finos.query.ungroup:{[t]
+  strCols:exec c from meta t where t="C";
+  $[count strCols
```

Comment:
ungroup in general is quite slow. It's faster to expand the "atomic" (including string) columns and raze the nested columns. There are probably edge cases that this misses but I think this would be a useful addition, maybe try the quick method in an error trap first and then go to the xungroup method if the quick method fails? In example below quickungroup is about 5x faster, and handles strings

quickungroup:{
 // pull out the grouped columns and the other columns
 groupcols:exec c from m:meta[x] where t=upper t,t<>"C";
 othercols:cols[x] except groupcols;
 // return if nothing to ungroup
 if[0=count groupcols; :x];
 // raze each of the groupcols
 // expand out the othercols, replicating rows for each grouped column
 // join together and reorder
 cols[x] xcols ((othercols#x) where count each x first groupcols),'flip groupcols!raze each x groupcols}
 
// tests
t1:([]sym:`a`b`c;bids:(1 2 3;4 5;enlist 6);asks:(8 7 6;5 4;enlist 3);notes:("here";"are";"strings")) 500000?3

\ts ungroup delete notes from t1
\ts quickungroup t1

Reply:
I think that the use of "meta" expresses the intent to the casual reader.
I'm not a fan of the error trap idea because Kx has been known to break functions in .Q / .q accidentally.
I do like the idea of a faster implementation, though!

Are you ok with the function signatures?
If so, what say we let this get merged into master and then you submit a follow up pull request with faster innards?
Thanks!

Reply:
Sounds good to me!